FCCV Docs (#15598)
Lightning-AI/lightning: docs/source-pytorch/data/custom_data_iterables.rst

.. _dataiters:

##################################
Injecting 3rd Party Data Iterables
##################################

When training a model with on a specific task, dataloading and preprocessing might become a bottleneck.
Lightning does not enforce a specific data loading approach nor does it try to control it.
The only assumption Lightning makes is that the data is returned as an iterable of batches.

For PyTorch-based programs these iterables are typically instances of :class:`~torch.utils.data.DataLoader`.

However, Lightning also supports other data types such as plain list of batches, generators or other custom iterables.

.. code-block:: python

    # random list of batches
    data = [(torch.rand(32, 3, 32, 32), torch.randint(0, 10, (32,))) for _ in range(100)]
    model = LitClassifier()
    trainer = Trainer()
    trainer.fit(model, data)

Examples for custom iterables include `NVIDIA DALI <https://github.com/NVIDIA/DALI>`__ or `FFCV <https://github.com/libffcv/ffcv>`__ for computer vision.
Both libraries offer support for custom data loading and preprocessing (also hardware accelerated) and can be used with Lightning.

> For example taking the example from FFCV's readme, we can use it with Lightning by just replacing the hardcoded ``ToDevice(0)``

Author: I agree, but on the other hand, it doesn't really matter if this is done by us or ffcv, right? If they do it, ours is a no-op. I wanted to stick to their example as close as possible

> For example taking the example from FFCV's readme, we can use it with Lightning by just replacing the hardcoded ``ToDevice(0)``

Reviewer: I would remove it from this example that doesn't use it but mention how it would be used if needed. It's cool to show that it can be removed with PL

which would always take the first GPU, no matter the actual process, with ``ToDevice(self.trainer.local_rank)`` to correctly map to the desired GPU.

.. code-block:: python

    from ffcv.loader import Loader, OrderOption
    from ffcv.transforms import ToTensor, ToDevice, ToTorchImage, Cutout
    from ffcv.fields.decoders import IntDecoder, RandomResizedCropRGBImageDecoder

    class CustomClf(LitClassifier):

>         def train_dataloader(self):

Author: I didn't mention it, since people are interested in how to integrate ffcv with PL not what requirements ffcv has and specifically as this is not related to training. Converting the dataset is a one-time thing and could be done separately.

>     class CustomClf(LitClassifier):

Reviewer 2: ```suggestion
    class CustomClassifier(LitClassifier):
```

                @pipeline_def(num_threads=4, device_id=self.trainer.local_rank)
                def get_dali_pipeline():
                    images, labels = fn.readers.file(file_root=images_dir, random_shuffle=True, name="Reader")
                    # decode data on the GPU
                    images = fn.decoders.image_random_crop(images, device="mixed", output_type=types.RGB)
                    # the rest of processing happens on the GPU as well
                    images = fn.resize(images, resize_x=256, resize_y=256)
                    images = fn.crop_mirror_normalize(
                        images,
                        crop_h=224,
                        crop_w=224,
                        mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
                        std=[0.229 * 255, 0.224 * 255, 0.225 * 255],
                        mirror=fn.random.coin_flip(),
                    )
                    return images, labels

                train_data = DALIGenericIterator(
                    [get_dali_pipeline(batch_size=16)],
                    ["data", "label"],
                    reader_name="Reader",
                )

                return train_data

> Lightning works seamlessly with all kinds of custom data iterables,

Reviewer 2: I don't think I fully understand this paragraph, can we make it more explicit and simpler to parse?

Author: @lantiga I revisited this paragraph, is it better now?

                    )
                    return images, labels

                train_data = DALIGenericIterator(
                    [get_dali_pipeline(batch_size=16)],
                    ["data", "label"],
                    reader_name="Reader",
                )

                return train_data

Limitations
------------
Lightning works with all kinds of custom data iterables as shown above. There are, however, a few features that cannot
be supported this way. These restrictions come from the fact that for their support,
Lightning needs to know a lot on the internals of these iterables.

- In a distributed multi-GPU setting (ddp),
  Lightning automatically replaces the DataLoader's sampler with its distributed counterpart.
  This makes sure that each GPU sees a different part of the dataset.
  As sampling can be implemented in arbitrary ways with custom iterables,
  there is no way for Lightning to know, how to replace the sampler.

- When training fails for some reason, Lightning is able to extract all of the relevant data from the model,
  optimizers, trainer and dataloader to resume it at the exact same batch it crashed.

>   This feature is called fault-tolerance and is limited to PyTorch DataLoaders as well as

Reviewer 2: ```suggestion
  This feature is called fault-tolerance and is limited to PyTorch DataLoaders.
```

>   Lighning also needs to know a lot about sampling, fast forwarding and random number handling to enable this,

Reviewer 2: ```suggestion
  Lighning needs to know a lot about sampling, fast forwarding and random number handling to enable fault tolerance,
```

>   meaning that this cannot be supported for arbitrary iterables either.

Reviewer 2: ```suggestion
  meaning that it cannot be supported for arbitrary iterables.
```
